[New blog post] continuous_batching (#3176)
huggingface/blog: continuous_batching.md

---
title: "Continuous batching from first principles" 
thumbnail: /blog/assets/continuous_batching/thumbnail.png
authors:
- user: ror
- user: ArthurZ
- user: mcpotato
---

# Continuous batching

![Title card](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/banner.png)

> ## Introduction

Reviewer: ```suggestion
```

---
title: "Continuous batching from first principles" 
thumbnail: /blog/assets/continuous_batching/thumbnail.png
authors:
- user: ror
- user: ArthurZ
- user: mcpotato
---

# Continuous batching

![Title card](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/banner.png)

> ## Introduction

Reviewer 2: ```suggestion
```
no need for this imo

Author: Removed it, thanks

If you've ever used Qwen, Claude, or any other AI chatbot, you've probably noticed something: it takes a while for the first word of the response to appear, and then words appear one-by-one on your screen with (hopefully) a regular and fast-paced frequency. That's because at the heart of it, all LLMs are just fancy next token predictors. First comes the **prefill** phase, where they ingest your initial prompt to predict one new token. Then they ingest that one as well, to produce another new token: this is the **decoding** phase. The process is repeated until they feel like they have generated enough.

> This generation process is computationally expensive: it requires processing billions of parameters for each token generated. To make these models practical for real-world applications, particularly when serving many users simultaneously, researchers and engineers have developed a range of efficient inference techniques.  

Author: Indeed, thanks!

One of the most impactful optimizations is **continuous batching**, which is the focus of this article. To understand how continuous batching works and why it's so effective, we'll need to build up from the fundamentals of how LLMs process tokens.

## Attention

> In order to understand continuous batching, it’s useful to have a clear understanding of the attention mechanism. In a Llama-like LLM, the neural network processes text by breaking it down into pieces that we call tokens, which are represented by vectors. We can think of tokens generally representing a single, simple word. For longer or rare words, several tokens may be used. For each token, the network computes a prediction of what the next token should be. It does this by applying a sequence of operations to each token.

Reviewer 3: ```suggestion
The attention mechanism is the central piece of how LLMs work. A language model processes text by breaking it down into pieces that we call tokens. We can conceptually think of "tokens" as "words", but sometimes a word might be composed of several tokens. For each token sequence, the network computes a prediction of what the next token should be.
```

(`llama-like` might be confusing for people not familiar with model ancestry 😂)
Also trying to defer the vector embeddings concept to some later point, otherwise I feel it's a piece of info readers will have to hold on to until it's used.

Most operations in the network are **token-wise**: each token is processed independently, and the output for a given token depends only on that token's content, not on any other tokens in the sequence. Operations like this include layer normalization or matrix multiplication. However, to create connections between words in a sentence, we need some operations where tokens can influence each other. 

> This is where attention comes in. **Attention layers are the only place where different tokens interact with each other**, something that even seasoned ML practitioners sometimes overlook. Understanding how a network connects tokens together means understanding attention.

Author: Indeed, it might be a little too much

Let's see how this works in practice. To explain the basics of attention, we'll focus on the case where we have one input prompt, which is the same as saying batch size is 1.

> Consider the initial prompt "I am sure this project". It is tokenized as 7 tokens: `[<bos>, I, am, sure, this, pro, ject]`. The `<bos>` token is a special token that is added at the start of the prompt (BoS stands for Beginning of Sequence).  

Reviewer 3: ```suggestion
Consider the initial prompt `I am sure this project`, tokenized as 7 tokens: `[<bos>, I, am, sure, this, pro, ject]`. The `<bos>`, or "Beginning of Sequence", is a special token we add at the start of the prompt to tell the language model that a new conversation starts here.
```

> Those incoming tokens form a tensor we call $x$, with shape $\left[1, n , d \right]$. This is because batch size is $1$, $n$ is the number of tokens and $d$ is the hidden dimension. 

Reviewer 2: these (and rest of the LaTeX too!) aren't rendered properly, just checked on hf.co/new-blog

syntax guide says: 

`You can mix latex with markdown. We pick $$YOUR LATEX$$ (for display) and \\(YOUR LATEX\\) (for inline) as delimiters. (example: \\( a=b+c \\) is my equation)`

<img width="872" height="395" alt="Screenshot 2025-11-25 at 11 48 06" src="https://github.com/user-attachments/assets/e5f29442-3a30-4b7f-92c0-cdbc1ae0edba" />

> Those incoming tokens form a tensor we call \\( x \\), with shape \\( \left[1, n , d \right] \\). This is because batch size is \\( 1 \\), \\( n \\) is the number of tokens and \\( d \\) is the hidden dimension. 

Reviewer 3: ```suggestion
Each token is represented inside the network with a vector of length `d` (the _hidden dimension_). Therefore, the seven incoming tokens form a tensor \\(x\\) with shape \\(\left[1, 7, d \right]\\). `1` is the number of sequences, or batch size, which is just one in our case. `7` is the sequence length, and `d` is the hidden dimension, or the size of each token representation. Going forward, we'll use \\(n\\) instead of `7` as the sequence length.
 ```

> Those tokens \\( x \\) are then projected by three matrices: the query projection \\( W_q \\),  the key projection \\( W_k \\) and the value projection \\( W_v \\). This produces three tensors \\( Q \\), \\( K \\) and \\( V \\), all of shape \\( \left[1, n , A \right] \\) where \\( A \\) is the dimension of an attention head. We call them respectively the **query, key and value states.** This is represented on the left in the figure below.

Reviewer 3: ```suggestion
Input tensor \\( x \\) is then projected by three matrices: the query projection \\( W_q \\),  the key projection \\( W_k \\) and the value projection \\( W_v \\). This produces three tensors \\( Q \\), \\( K \\) and \\( V \\), all of shape \\( \left[1, n , A \right] \\), where \\( A \\) is the dimension of the attention head. We call them the **query, key and value states,** respectively. This is represented on the left in the figure below.
```

![proj_and_mul.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/proj_and_mul.png)

Next, tensors $Q$ and $K$ are multiplied together to measure similarity between tokens, producing a tensor of shape $\left[ 1, n , n \right]$. This is why we say that attention has quadratic complexity in sequence length. Computing $QK^T$ requires $\mathcal{O} \left( n^2 d \right)$ operations, so the cost is a square of $n$ the sequence length. It is represented on the right in the figure above.

We then apply a boolean **attention mask** to $QK^T$ to control which tokens can interact, as represented in the figure below. In this figure, the attention mask is a **causal mask**, meaning each token only interacts with tokens that came before it. This follows the intuition that a cause must come before its consequence, hence the name causal mask. The attention mask is crucial because it dictates all token interactions in the network. **Set all attention mask values to False and no token will ever interact with another in the whole network.** We'll examine attention masks more closely in a few paragraphs.

> ![masking_and_softmax.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/masking_and_softmax.png)

Reviewer 2: not a must but I'd love to see the tokens on the left hand side of the matrices

> Finally, after applying the attention mask, we take a row-wise softmax and multiply the result by the value projection $V$ to get the output of one attention head, of shape $\left[ 1, n , A \right]$. We offer a visual summary of the whole process in the following figure.

Reviewer 2: nit: would be nice to say token-wise softmax to be coherent

![attention.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/attention.png)

> 

Reviewer 3: ```suggestion

The process we just described, where we take an input sequence, pass it through multiple attention layers and compute a score for the next token, is called "prefill". This is because, as we'll see in a moment, much of the computation we performed can be cached and reused – hence, we are _prefilling_ the cache. Thanks to the use of the cache, sequence generation can proceed from this moment on by handling one token at a time and expanding the cache, making successive tokens much faster to generate than the initial full-sequence computation.
```

Trying to provide some breadcrumbs and introducing prefill here, I hope it makes sense 🤞

Author: I moved the breadcrumbs a bit but kept them, thanks!

![cant_see_me.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/cant_see_me.png)

This follows the idea of the causal mask: since "*will*" comes after all previous tokens, it does not change their attention calculation.
For text, causal attention is by far the most common. We will focus on that case from now on. But non-causal attention also exists, especially when dealing with images.
Considering we only need the next-token prediction for the "*will*" token, we can simplify the attention mechanism by only computing the output for this token.

Moreover, we already computed the $K$ and $V$ states for the tokens "*\<bos\>*", … ,  "*ject*" during the previous forward pass: if they have been stored, we do not need to recompute them again. This is the **KV cache**: the list of key and value states created during generation. It essentially allows one to reduce the compute cost of generating token $n+1$ from $\mathcal{O} \left( n^2 \right)$ to $\mathcal{O} \left( n \right)$ by avoiding recomputation of key and value projections, while paying a memory cost of  $\mathcal{O} \left( n \right)$.

> ![kv_cache.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/kv_cache.png)

Reviewer 3: Somewhere in this section we could potentially link to https://huggingface.co/blog/not-lain/kv-caching or https://huggingface.co/blog/kv-cache for people that want to dive deeper or see other visualizations.

In the figure above, only the tokens in white are computed: instead of computing the keys and values for 8 tokens, we compute them for 1. You can see that through KV caching, a lot of compute is saved.

> Let's be a bit more specific around the cache size, because it's a good opportunity to tour the shapes present in our model. For a model with \\( \mathcal L \\) attention layers and \\( H \\) attention heads with head dimension \\( A \\), the total cache size needed to store one token will be \\( 2 *\mathcal L * AH \\) with a factor of \\( 2 \\) to account for both \\( K \\) and \\( V \\).  

Reviewer 3: ```suggestion
Let's be a bit more specific about the cache size, because it's a good opportunity to examine the shapes present in our model. For a model with \\( \mathcal L \\) attention layers and \\( H \\) attention heads with head dimension \\( A \\), the total cache size needed to store one token will be \\( 2 *\mathcal L * AH \\) with a factor of \\( 2 \\) to account for both \\( K \\) and \\( V \\).  
```

Let's pretend that the available memory is very constrained, and that we can only pass $m=4$ tokens per forward pass. If we have an initial prompt with $n = 7$ tokens, we need to split it in $\lceil n /m \rceil = 2$ chunks (rounding up 7/4 = 1.75 to 2). We illustrate the example below using the same $n$ and $m$ notations:

![chunked prefill.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/chunked_prefill.png)

We can do that thanks to the KV cache. We store the KV states during the first prefill split, and during the second prefill split, we concatenate them left of the new KV states. We also adapt the attention mask accordingly. Visually, it looks like we split the non-chunked prefill in the middle.

The key insight: cached KV states let us process the prompt incrementally without losing information.

Although we showed here an example where we split the prefill into 2 chunks, chunked prefill can be used to split the prefill in any way we want, adapting flexibly to memory constraints.  
You should by now be equipped with the tools to understand Continuous Batching.

## Continuous batching

> So far, we have only treated the case of batch size one, i.e. we only generate tokens for one prompt at a time. But in the context of evaluation or model serving, we want to generate tokens for a large number of prompts. To increase the **throughput**, which is the number of generated tokens divided by the time it took to generate them, the best course of action is to generate tokens for a batch of several prompts.

Author: Not sure about this one, it feels like it's just a number of token. How do you feel about "the number of tokens generated per second" ?

> So far, we have only treated the case of batch size one, i.e. we only generate tokens for one prompt at a time. But in the context of evaluation or model serving, we want to generate tokens for a large number of prompts. To increase the **throughput**, which is the number of generated tokens divided by the time it took to generate them, the best course of action is to generate tokens for a batch of several prompts.

Reviewer 2: yep sounds good!

> To batch prompts together, the naive way is to add an axis to both input tensors, which will be the batch axis. This way we can pass two prompts and two attention masks, one for each. However, this comes with a constraint on the shape of the inputs: we need all prompts to have the same length (since tensors must be rectangular). To achieve this, we usually add padding on the left so the new token prediction always comes from the rightmost token. We also modify the attention mask of each prompt accordingly. This is shown below:

Reviewer 3: ```suggestion
To batch prompts together, the naive way is to add an axis to both input tensors: token sequence and attention mask. However, this comes with a constraint on the shape of the inputs: we need all prompts to have the same length, because tensors must be rectangular. To achieve this, we usually add padding on the left so the new token prediction always comes from the rightmost token. We also modify the attention mask of each prompt accordingly, as shown below:
```

![padding.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/padding.png)

where the padding tokens `<pad>` are coloured in orange. Then we can perform the forward pass as we used to, with the added dimension of the batch size. This is called **batched generation**: efficient for same-length prompts, but wasteful when lengths vary. It is illustrated below:

> ![batched_generation.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/batched_generation.png)

Author: Not confused, you are right! Thanks for the catch.

where `<eos>` means "End Of Sequence", this is a special token to indicate the model has reached the end of generation.

> The drawback of batched generation is that if one prompt finishes generation before the other by generating an `<eos>` token, any further generated token is useless. And this goes on until the longest request of the batch finishes. Of course, we can remove the prompts that have reached an `<eos>` token from the batch and save some compute and memory, but saving resources is not the goal here: throughput is. 

Reviewer 3: ```suggestion
The drawback of batched generation is that if one prompt finishes generation before the other one by generating an `<eos>` token, all further generated token are useless. And this goes on until the longest request of the batch finishes. Of course, we can remove the prompts that have reached an `<eos>` token from the batch and save some compute and memory, but saving resources is not the goal here: throughput is. 
```

Instead of just removing the finished prompt from the batch, we can replace it with a prompt that's waiting for generation. We will call this **dynamic scheduling**, but it is also called dynamic batching. Dynamic scheduling is great to maintain throughput while ensuring any token generated by a forward pass is relevant. But because of the way we batched prompts together, it has a major drawback: we need a lot of padding when swapping prompts.

> ![dynamic_batching.png](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/continuous_batching/dynamic_batching.png)

Reviewer 3: Maybe point out that the newly-inserted sequence needs to go through a prefill phase while the others are decoding one token at a time, so we need to add as many tokens (per sequence) as the new sequence length to make shapes match.

Author: Good point, done
